### PR TITLE
945 add beam modulator

### PR DESCRIPTION
**Pull request by Derstilon**
#732 #945 

To be tested with https://github.com/yaptide/yaptide/pull/449

### CODE REVIEW

**Comment by grzanka** [comment]
@Derstilon increase the timeout on tests

**Comment by grzanka** [comment]
Does it work ?

I get following error:

```
ERROR
orderList is not iterable
TypeError: orderList is not iterable
    at orderAccordingToList (https://grzanka-special-giggle-r5gw9r9qvjfxqwx-3000.preview.app.github.dev/static/js/bundle.js:35937:30)
    at recreateOrderInEstimators (https://grzanka-special-giggle-r5gw9r9qvjfxqwx-3000.preview.app.github.dev/static/js/bundle.js:34229:74)
    at recreateRefsInResults (https://grzanka-special-giggle-r5gw9r9qvjfxqwx-3000.preview.app.github.dev/static/js/bundle.js:34252:29)
    at ./src/ThreeEditor/examples/examples.ts (https://grzanka-special-giggle-r5gw9r9qvjfxqwx-3000.preview.app.github.dev/static/js/bundle.js:17113:178)
    at options.factory (https://grzanka-special-giggle-r5gw9r9qvjfxqwx-3000.preview.app.github.dev/static/js/bundle.js:265127:31)
    at __webpack_require__ (https://grzanka-special-giggle-r5gw9r9qvjfxqwx-3000.preview.app.github.dev/static/js/bundle.js:264533:33)
    at fn (https://grzanka-special-giggle-r5gw9r9qvjfxqwx-3000.preview.app.github.dev/static/js/bundle.js:264784:21)
    at ./src/ThreeEditor/components/Dialog/OpenFileDialog.tsx (https://grzanka-special-giggle-r5gw9r9qvjfxqwx-3000.preview.app.github.dev/static/js/bundle.js:8464:76)
    at options.factory (https://grzanka-special-giggle-r5gw9r9qvjfxqwx-3000.preview.app.github.dev/static/js/bundle.js:265127:31)
    at __webpack_require__ (https://grzanka-special-giggle-r5gw9r9qvjfxqwx-3000.preview.app.github.dev/static/js/bundle.js:264533:33)
```

image

**Comment by grzanka** [comment]
@Derstilon is it ready for review ?

**Comment by grzanka** [comment]
@Derstilon increase timeout for tests

image

**Comment by grzanka** [comment]
@Derstilon the time constrains for the tests are defined here:
https://github.com/yaptide/ui/blob/945-add-beam-modulator/src/__tests__/ShieldhitConverter.test.ts#L144

this is the file you should modify

**Review by ostatni5 — COMMENTED** [review]
image
Add looks more like a button than a label.
The old design was better 
image

File: src/reset.d.ts
Comment (ostatni5) [inline src/reset.d.ts:1]: 🎧

File: src/services/AuthService.tsx [hunk 1]
```diff
@@ -25,12 +25,35 @@ export interface AuthProps {
...
+	);
+};
+const isYaptideResponse = (obj: unknown): obj is YaptideResponse => {
+	return (
+		typeof obj === 'object' &&
+		obj !== null &&
+		'message' in obj &&
+		typeof obj.message === 'string'
+	);
+};
```
Comment (ostatni5) [inline src/services/AuthService.tsx:43]: Add guard for has fields. zrob cos mondrego  🔢

File: src/types/SimulationProperties.ts [hunk 2]
```diff
@@ -1,15 +1,20 @@
...
+ * isSimulationElement is always true and is defined in the type
+ * because you can't implement a type in a class without defining at least one property.
  * all other properties are optional and are used to properly handle the object in the editor.
  */
 export type SimulationPropertiesType = {
-	readonly isSimulationElement?: true;
+	//required
+	readonly isSimulationElement: true;
+
+	//optional
```
Comment (ostatni5) [inline src/types/SimulationProperties.ts:11]:
```suggestion
```

File: src/types/SimulationProperties.ts [hunk 3]
```diff
@@ -1,15 +1,20 @@
 /**
  * Type for properties of a simulation object.
- * isObject3D is always true and is defined in the type because you can't implement a type in a class without defining at least one property.
+ * isSimulationElement is always true and is defined in the type
+ * because you can't implement a type in a class without defining at least one property.
  * all other properties are optional and are used to properly handle the object in the editor.
  */
 export type SimulationPropertiesType = {
-	readonly isSimulationElement?: true;
+	//required
```
Comment (ostatni5) [inline src/types/SimulationProperties.ts:8]:
```suggestion
```

File: src/ThreeEditor/Simulation/Base/SimulationZone.ts [hunk 4]
```diff
@@ -138,15 +160,28 @@ export abstract class SimulationZone
 		return this;
 	}
 
-	fromJSON(json: SimulationZoneJSON): this {
-		const { uuid, name, materialUuid, materialPropertiesOverrides: overrides } = json;
+	reconstructMaterialFromJSON(json: SimulationZoneJSON) {
```
[the remark this answers is not in the document]
Comment (Derstilon) [reply src/ThreeEditor/Simulation/Base/SimulationZone.ts:163]: done and moved near other private methods

File: src/ThreeEditor/Simulation/SpecialComponents/BeamModulator.ts [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+		}
+		this.geometry.computeBoundingSphere();
+		this.editor.signals.objectSelected.dispatch(this);
+	}
+	get geometryParameters(): ModulatorParameters {
+		return {
+			zoneUuid: this._zoneUuid
+		};
+	}
+	set geometryParameters(parameters: ModulatorParameters) {
```
[the remark this answers is not in the document]
Comment (Derstilon) [reply src/ThreeEditor/Simulation/SpecialComponents/BeamModulator.ts:76]: https://github.com/yaptide/ui/issues/1088

File: src/ThreeEditor/components/Editor/EditorAppBar/EditorAppBar.tsx [hunk 6]
```diff
@@ -82,6 +82,7 @@ function EditorAppBar({ editor }: AppBarProps) {
 
 	const openFile = (files: FileList) => {
 		if (editor) editor.loader.loadFiles(files);
+		//TODO: rewrite to support our versioning and types of data. Default loader is now mostly useless
```
[the remark this answers is not in the document]
Comment (Derstilon) [reply src/ThreeEditor/components/Editor/EditorAppBar/EditorAppBar.tsx:85]: https://github.com/yaptide/ui/issues/1089

**Review by ostatni5 — APPROVED** [review]
image

**Pull request merged**